src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -208,37 +211,58 @@ maxPacketSize = 256 * 2 ^ (20 :: Int) - 128 * 2 ^ (10 :: Int)
 -- @since 0.1.0.0
 makePacketSender ::
   MonadUnliftIO m =>
-  -- | The maximum packet payload size. This must be a 'Word32' value greater 
+  -- | The maximum packet payload size. This must be a 'Word32' value greater
   -- than or equal to 'minPacketSize' and less than 'maxPacketSize'. Otherwise,
   -- a 'PacketSizeError' exception will be thrown.
   Word32 ->
+  -- | Optional rate limit for publishing as bytes per second.
+  Maybe Word32 ->
   (ByteString -> m ()) ->
   ByteString ->
   m ()
-makePacketSender limit publish message = do
-  unless (isValidPacketSize limit) do
-    throwIO (PacketSizeError $ toInteger limit)
+makePacketSender packetSizeLimit mRateLimit publish message = do
+  unless (isValidPacketSize packetSizeLimit) do
+    throwIO (PacketSizeError $ toInteger packetSizeLimit)
 
   if fromIntegral (ByteString.length message) <= maxPayloadSize
     then do
-      -- Perform a single publish on the current thread in case that the 
+      -- Perform a single publish on the current thread in case that the
       -- @message@ length is less than the maximum payload size.
       let packet = Packet message 0 1
       publish (wireWrapPacket' packet)
-    else do
-      caps <- liftIO getNumCapabilities
-      jobs <- newTVarIO 0
-
-      replicateConcurrently_ caps $ fix \next ->
-        atomically (takeJobId jobs) >>= \case
-          Nothing -> pure ()
-          Just jobid -> do
-            let packet = makePacket jobid
-            publish (wireWrapPacket' packet)
-            next
+    else case mRateLimit of
+      Nothing -> do
+        caps <- liftIO getNumCapabilities
+        jobs <- newTVarIO 0
+
+        replicateConcurrently_ caps $ fix \next ->
+          atomically (takeJobId jobs) >>= \case
+            Nothing -> pure ()
+            Just jobid -> do
+              let packet = makePacket jobid
+              publish (wireWrapPacket' packet)
+              next
+      Just rateLimit -> do
+        -- Throughput is rate limited on a 1 second interval and we don't subdivide packets to comply
+        -- with the rate limit, so the payload size must be less than or equal to the rate limit.
+        -- NB: Since `maxPayloadSize` is already at least 1, this ensures that `rateLimit` is not 0.
+        unless (maxPayloadSize <= rateLimit) $
```

Comment:
Can you push this further upstream? e.g if `rateLimit` were a `Natural` then we wouldn't need this check.

Reply:
Ah, I thought that might be the case but like you said I also think it's still a better fit if we never need the value to be negative or zero.